File: src/main/java/io/github/binaryfoo/crypto/CaPublicKeyTable.java
package io.github.binaryfoo.crypto;

import io.github.binaryfoo.io.ClasspathIO;
import org.apache.commons.io.IOUtils;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;

public class CaPublicKeyTable {

    private static final List<CaPublicKey> keys = new ArrayList<>();
    static {
        for (String line : ClasspathIO.readLines("ca-public-keys.txt")) {
            String[] fields = line.split("\\t");
            String exponent = fields[1];
            String index = fields[2];
            String aid = fields[3];
            String modulus = fields[4];
            keys.add(new CaPublicKey(aid, index, exponent, modulus));
        }
    }

    public static CaPublicKey getEntry(String aid, String index) {
        for (CaPublicKey key : keys) {
            if (key.getAid().equals(aid) && key.getIndex().equals(index)) {
                return key;
            }
        }
        return null;
    }
}

File: src/main/java/io/github/binaryfoo/crypto/CaPublicKeyTable.kt
package io.github.binaryfoo.crypto

import io.github.binaryfoo.res.ClasspathIO

/**
 * Table of trust anchors.
 *
 * From https://www.eftlab.co.uk/index.php/site-map/knowledge-base/243-ca-public-keys
 */
class CaPublicKeyTable {
  companion object {

    val keys: List<CaPublicKey> = ClasspathIO.readLines("ca-public-keys.txt").map { line ->
      try {
        val fields = line.split(Regex("\\t"))
        val exponent = fields[1]
        val index = fields[2]
        val rid = fields[3]
        val modulus = fields[4]
        CaPublicKey(rid, index, exponent, modulus)
      } catch (e: Exception) {
        throw RuntimeException("Failed reading CA public key: $line", e)
      }
    }

    @JvmStatic
    fun getEntry(aid: String, index: String): CaPublicKey? = keys.firstOrNull { it.rid == aid && it.index == index }
  }
}
